feat(mobile): Manual asset upload

shenlong-tanwen opened the pull request:
In this PR, I've added a basic support for #1684

### Changes made in the PR
- User can use the upload button from the Asset viewer to manually upload the single asset to the server or they can select **_at-most 30_** assets from the timeline

<p align="center">
<img src="https://github.com/immich-app/immich/assets/139912620/00e1a1ce-0d6d-4be7-a5e4-c3f50781c0d0 " alt="Upload Asset Example" height="300">
<img src="https://github.com/immich-app/immich/assets/139912620/b0d7efec-a77f-4474-9e29-c0695d7c3918" alt="Upload Asset from Timeline Example" height="300">
</p>

- DraggableScrollbarSheet in homepage is made dynamic to show actions related to the type of assets selected. If it contains only local assets, only the Share, Delete and Upload are displayed. If it contains both local and remote assets, favorite, add to album and other similar options are displayed.

<strong>Local Only Selection</strong>
<p align="center">
<img src="https://github.com/immich-app/immich/assets/139912620/c7aa3298-bac4-4631-b9cb-b0e08e61ccf7" alt="Bottom Sheet Local Only Example" height="300">
</p>
<strong>Local + Remote Selection</strong>
<p align="center">
<img src="https://github.com/immich-app/immich/assets/139912620/6309fba2-1a7f-4f3d-a497-a8300785c955" alt="Upload Asset from Timeline Example" height="300">
</p>

- While testing, It seemed extremely hard to trigger force refresh from the homepage. Going through the code, found that the current interval between two successive refresh to force it was set to 2 seconds. I've bumped it to 4 seconds to make it a little bit more easier to trigger force refresh if required.

### How Has This Been Tested?
✅ Multi asset upload from homepage
✅ Single asset upload from asset view

### Potential Improvements
- Currently, Favorite and other options to be performed on Remote only assets from the bottom sheet in the homepage are ignoring the local assets with a toast stating that they are not supported. The user can be presented with a prompt to upload them before performing the required action.
- Local Notification handling for Asset upload progress, failure and error instead of showing a Toast
- Need to analyze the performance impact of the app because of the change to make the homepage bottom sheet dynamic. From my basic testing, it did not seem to impact much in Performance. But it would be best to test it out once with large number of asset selections.

shenlong-tanwen committed:
fix: exclude albums filter in backup provider

shenlong-tanwen committed:
refactor: Separate builder methods for Top Control App Bar buttons

shenlong-tanwen committed:
fix: Show download button only for Remote only assets

shenlong-tanwen committed:
fix(mobile): Force Refresh duration is too low to trigger it consistently

shenlong-tanwen committed:
feat(mobile): Make Buttons dynamic in Home Selection DraggableScrollableSheet

shenlong-tanwen committed:
feat(mobile): Manual Asset upload

alextran1502 commented:
Hello, what is the limitation of 30 uploads at once?

shalong-tanwen commented:
> Hello, what is the limitation of 30 uploads at once?

It is a not a limitation with the stack but rather one set by us to easily handle cases such as:
- App going into the background and thereby possibly killing off the upload in-between
- Platform specific battery optimisation killing the app, etc.

I just felt it would be easier to handle such cases by asking the user to upload the assets in batches so that we can be sure that all the assets will be uploaded before the app is killed prematurely. Because with the current implementation, one can upload assets even without background / foreground sync enabled.

We can remove the hardlimit, but we might need to improve the current implementation to make it resilient to above cases.

Even with the hardlimit, it might be possible that the upload might be interrupted if a user is uploading a huge file and the apps gets terminated somehow.

martyfuhry commented on mobile/lib/modules/home/ui/control_bottom_app_bar.dart:
```diff
@@ -66,19 +76,35 @@ class ControlBottomAppBar extends ConsumerWidget {
     return DraggableScrollableSheet(
-      initialChildSize: 0.30,
+      initialChildSize: hasRemote ? 0.30 : 0.15,
       minChildSize: 0.15,
-      maxChildSize: 0.57,
+      maxChildSize: hasRemote ? 0.57 : 0.15,
```
Why are these changed?

shalong-tanwen replied:
Those are for the `DraggableScrollbarSheet` related change where if all the selections are only local assets, we shouldn't let the user drag the bottom sheet and only present them with the options available for local only assets. Setting the initial and maximum values to the same seemed like an easier way to achieve it

martyfuhry commented on mobile/lib/modules/home/views/home_page.dart:
```diff
@@ -35,6 +36,7 @@ class HomePage extends HookConsumerWidget {
   Widget build(BuildContext context, WidgetRef ref) {
     final multiselectEnabled = ref.watch(multiselectProvider.notifier);
     final selectionEnabledHook = useState(false);
+    final selectionMode = useState(AssetState.remote);
```
Hmm, maybe I'd call this something else like `selectionAssetState`?

shalong-tanwen replied:
Sounds better. Made the change and refactored the method to a variable in the process.

martyfuhry reviewed (commented):
Looks cool! I didn't pull it down to try it out, but I left some initial thoughts above. I'm not super sure what the use case for this is, since I use the auto backup feature. I suppose there could be some confusion about uploading images manually versus automatically with this feature, but I doubt it's a big deal.

shenlong-tanwen committed:
refactor(mobile): Replace _showToast with ImmichToast calls

shenlong-tanwen committed:
refactor(mobile): home_page selectionAssetState handling

shenlong-tanwen committed:
chore(mobile): min and initial size of DraggableScrollState increased

This is to prevent the buttons in the bottom sheet getting clipped behind the 3 way navigation buttons
in the default density of Android devices

shalong-tanwen commented on mobile/lib/modules/backup/providers/backup.provider.dart:
```diff
@@ -594,7 +615,7 @@ class BackupNotifier extends StateNotifier<BackUpState> {
     final List<BackupAlbum> excludedBackupAlbums = await _db.backupAlbums
         .filter()
-        .selectionEqualTo(BackupSelection.select)
+        .selectionEqualTo(BackupSelection.exclude)
```
From what I understood, the piece of code below the change is used to update the `lastBackup` time of albums by updating the currently selected and excluded albums from the current state with the saved album meta details from the app db. The filter logic for fetching the selected and excluded albums from the db seemed wrong so I updated it. 

Gentle ping @fyfrey

fyfrey replied:
This change looks good. No idea how the wrong query ever made it into the code:-)

shalong-tanwen commented on mobile/lib/modules/backup/providers/backup.provider.dart:
```diff
@@ -656,6 +677,96 @@ class BackupNotifier extends StateNotifier<BackUpState> {
     }
   }
+
+  void _showToast(BuildContext context, String msg, ToastType toastType) {
```
I was skeptical about this change as well. Probably not worth the additional function. I've removed it now

waclaw66 commented:
> Looks cool! I didn't pull it down to try it out, but I left some initial thoughts above. I'm not super sure what the use case for this is, since I use the auto backup feature. I suppose there could be some confusion about uploading images manually versus automatically with this feature, but I doubt it's a big deal.

I'm happy for this feature. I don't/wan't to use automatic backup, manual selective upload of photos is my use case.

alextran1502 commented:
Hello, I've finished the first round of testing of this PR 

- We don't have an indication of upload progress when uploading from the asset detail view. It can be confusing to the user if they are on a slower network, and it takes some time to upload the asset. What do you think?
- After uploading a video, the state isn't updated correctly. Therefore you cannot upload other assets unless going to the backup page and tap on the Cancel button at the bottom of the page.

shenlong-tanwen committed:
feat(mobile): notifications for manual upload progress

shalong-tanwen commented:
Refactored the PR to handle upload progress update using local notifications. New dependency `flutter_local_notifications` is added to make it easier to handle platform specific local notifications. 

The notification works the same way as the change made in this PR - #781 except that the progress notifications are displayed always for manual uploads. Detailed notifications are displayed only when the user explicitly enables it under Notification settings. However, if a user is trying to upload only a single asset, we will display detailed notification for the single asset so as to provide better visibility when a user uploads a single large asset file. If the notification permission is not granted, the user will not be notified of the upload progress nor the final status of the upload. 

The older platform specific code for handling background service upload notification can be refactored to use this plugin in a separate PR. 

The other issue with App State not updating properly when uploading a video asset was due to us trying to show a toast when the context is unmounted. Since most part of them are refactored to use local notifications now, that issue should be fixed as well.

alextran1502 committed:
wording

waclaw66 commented:
Could you please make the limit much higher or rather remove it completely?
I think it makes no different when the app is killed after 31 or 300 uploads.
Look at following usecase as I described above. I don't want to use automatic backup for many reasons. I want to upload daily shots (e.g. 300) in the evening. Now I have to select batch of 30 manually and let them upload and do the same 9times again. It's exhausting. When 31+ assets is selected to upload it refuses to start.
My proposal is to remove that limit completely or if you don't agree with my proposal upload first 30 of all selected (300) assets at least. Then I would need to select all day and click to upload 10times instead of manually selecting 300 images manually. It would be less painfull.
Thanks in advance for considering the enhancement.

daniele-athome commented:
I second that. Battery optimizations don't kick in if there is a foreground service notification (on Android, I don't know about iOS though).
Random app kills due to low resources could happen anyway, I don't think we can blame the app for that (OOM is called on rare, very extreme circumstances).

shalong-tanwen commented:
> Could you please make the limit much higher or rather remove it completely? I think it makes no different when the app is killed after 31 or 300 uploads. 

We can remove the limit for now and can re-add it in future if there are reports of the app getting killed. I'll handle this in the near future or you guys can feel free to open a PR to remove the hard limit. It is handled in the `home_page.dart` file

https://github.com/immich-app/immich/blob/5d1011b482831feb0411c9fe0ae60be1c6146ff3/mobile/lib/modules/home/views/home_page.dart#L153-L159

banto6 commented:
Great feature, but currently the interaction is rather strange, will it be optimized for interaction in the future? For example, add an extra button to each album or homepage to realize the upload function

shenlong-tanwen commented:
> Great feature, but currently the interaction is rather strange, will it be optimized for interaction in the future? For example, add an extra button to each album or homepage to realize the upload function

I am planning to make this a bit more easier, yes. Adding an option to the local album page sounds like a great feature.

banto6 commented:
Sure.
